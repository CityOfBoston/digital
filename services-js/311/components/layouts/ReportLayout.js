// @flow
/* eslint react/no-unused-prop-types: 0 */

import React from 'react';

import Nav from '../common/Nav';
import LocationMap from '../location';
import StartFormContainer from '../report-start';
import ReportFormContainer from '../report-form';

import type { LoopbackGraphql } from '../../data/graphql/loopback-graphql';
import type { Service, ServiceSummary } from '../../data/types';

type ServiceProps = {
  view: 'service',
  code: string,
  service: ?Service,
  pickLocation: boolean,
};

type SummaryProps = {
  view: 'summaries',
  serviceSummaries: ?ServiceSummary[],
  pickLocation: boolean,
};

export type InitialProps = ServiceProps | SummaryProps;

export type Props = {
  initialProps: InitialProps,
  loopbackGraphql: LoopbackGraphql,
  showServiceForm: (ServiceSummary) => void,
  goToReportForm: () => void,
}

export default class ReportLayout extends React.Component {
  props: Props;

  static getTitle(initialProps: InitialProps) {
    switch (initialProps.view) {
      case 'summaries': return 'Report a Problem';
      case 'service': {
        const { service, pickLocation } = initialProps;

        if (pickLocation) {
          return 'Choose location';
        } else if (service) {
          return service.name;
        } else {
          return 'Not found';
        }
      }
      default: return '';
    }
  }

  render() {
    const { initialProps: { pickLocation }, goToReportForm } = this.props;

    return (
      <div>
        <Nav />
        <LocationMap active={pickLocation} goToReportForm={goToReportForm} />

        {this.renderContent()}
      </div>
    );
  }

  renderContent() {
    const { initialProps, showServiceForm, loopbackGraphql } = this.props;

    switch (initialProps.view) {
      case 'summaries':
        return (
          <StartFormContainer serviceSummaries={initialProps.serviceSummaries} showServiceForm={showServiceForm} />
        );

      case 'service': {
        const { service, pickLocation } = initialProps;
        if (pickLocation) {
          return null;
        } else {
          return <ReportFormContainer service={service} loopbackGraphql={loopbackGraphql} />;
        }
      }

      default:
        return null;
    }
  }
}
