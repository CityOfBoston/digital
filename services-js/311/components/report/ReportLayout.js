// @flow
/* global liveagent */
/* eslint jsx-a11y/anchor-has-content: 0 */

import React from 'react';
import { css } from 'glamor';
import Router from 'next/router';
import type { Context } from 'next';
import { action, computed } from 'mobx';
import { observer } from 'mobx-react';

import type { RequestAdditions } from '../../server/next-handlers';

import Nav from '../common/Nav';
import LocationMap from './map/LocationMap';
import type { MapMode } from './map/LocationMap';
import HomeDialog from './home/HomeDialog';
import RequestDialog from './request/RequestDialog';

import { MEDIA_LARGE, HEADER_HEIGHT } from '../style-constants';

import makeLoopbackGraphql from '../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../data/dao/loopback-graphql';

import loadTopServiceSummaries from '../../data/dao/load-top-service-summaries';
import loadService from '../../data/dao/load-service';

import type { Service, ServiceSummary } from '../../data/types';

import getStore from '../../data/store';
import type { AppStore } from '../../data/store';

type HomeData = {|
  view: 'home',
  topServiceSummaries: ServiceSummary[],
  stage: 'home' | 'choose',
|};

type RequestData = {|
  view: 'request',
  code: string,
  service: ?Service,
  stage: 'questions' | 'location' | 'contact' | 'submit',
|};

export type InitialProps = {|
  data: HomeData | RequestData,
|};

export type Props = {|
  store: AppStore,
  /* :: ...InitialProps, */
|};

// This is the main container for content. We want to be at least full height on
// large screens to push the footer down to where you need to scroll for it.
const CONTAINER_STYLE = css({
  minHeight: 0,
  position: 'relative',
  justifyContent: 'flex-start',
  [MEDIA_LARGE]: {
    minHeight: '100vh',
  },
});

const BACKGROUND_MAP_CONTAINER_STYLE = css({
  position: 'fixed',
  width: '100%',
  top: HEADER_HEIGHT,
  bottom: 0,
  background: '#9B9B9B',
});

// We have one class for picking the service type and doing the entire request
// so that we can keep a consistent Map background behind the
// dialog and not repeatedly re-mount it.
//
// This class has two "views": the "summaries" view that lets the user pick a
// service request, and the "service" view for the request flow. That flow
// proceeds through the "questions" "location" and "contact" stages.
@observer
export default class ReportLayout extends React.Component {
  props: Props;
  state: {
    locationMapActive: boolean,
  }
  loopbackGraphql: LoopbackGraphql;

  static async getInitialProps({ query, req, res }: Context<RequestAdditions>): Promise<InitialProps> {
    const loopbackGraphql = makeLoopbackGraphql(req);

    let data;

    if (query.code) {
      data = await ReportLayout.getRequestData(query, res, getStore(), loopbackGraphql);
    } else {
      data = await ReportLayout.getHomeData(query, loopbackGraphql);
    }

    return {
      data,
    };
  }

  static async getHomeData({ stage }, loopbackGraphql): Promise<HomeData> {
    const store = getStore();

    return {
      view: 'home',
      topServiceSummaries: store.topServiceSummaries || await loadTopServiceSummaries(loopbackGraphql),
      stage: stage === 'choose' ? stage : 'home',
    };
  }

  static async getRequestData({ code, stage }, res, store, loopbackGraphql): Promise<RequestData> {
    let service = store.serviceCache.get(code);

    if (!service) {
      service = await loadService(loopbackGraphql, code);

      if (!service && res) {
        res.statusCode = 404;
      }
    }

    stage = stage || 'questions';

    switch (stage) {
      case 'questions':
      case 'location':
      case 'contact':
      case 'submit':
        return {
          view: 'request',
          code,
          service,
          stage,
        };
      default:
        throw new Error(`Unknown stage: ${stage}`);
    }
  }

  // TODO(finneganh): Move service cache and lookup out of this class
  @action static addServiceToCache(store: AppStore, service: Service) {
    store.serviceCache.set(service.code, service);
  }

  constructor(props: Props) {
    super(props);

    this.loopbackGraphql = makeLoopbackGraphql();

    this.updateStoreWithProps(props);

    this.state = {
      locationMapActive: false,
    };
  }

  componentWillReceiveProps(props: Props) {
    this.updateStoreWithProps(props);
    this.setState({ locationMapActive: false });
  }

  startChat = () => {
    const { liveAgentButtonId } = this.props.store;
    liveagent.startChat(liveAgentButtonId);
  }

  @computed get mapActivationRatio(): number {
    const { store: { ui }, data: { view, stage } } = this.props;
    if (view === 'home' && stage === 'home') {
      return Math.min(1.0, ui.scrollY / (ui.visibleHeight * 0.75));
    } else {
      return 0;
    }
  }

  @action
  updateStoreWithProps(props: Props) {
    const { store, data } = props;

    switch (data.view) {
      case 'home':
        store.topServiceSummaries = data.topServiceSummaries;
        store.currentService = null;
        break;

      case 'request': {
        const { service } = data;
        if (service) {
          ReportLayout.addServiceToCache(store, service);
        }
        store.currentService = service;
        break;
      }

      default:
        break;
    }
  }

  routeToServiceForm = async (code: string, stage: string = 'questions') => {
    if (stage === 'questions') {
      await Router.push(`/report?code=${code}`, `/report/${code}`);
    } else {
      await Router.push(`/report?code=${code}&stage=${stage}`, `/report/${code}/${stage}`);
    }

    window.scrollTo(0, 0);
  }

  setLocationMapActive = (active: boolean) => {
    const { locationMapActive } = this.state;

    if (locationMapActive !== active) {
      this.setState({ locationMapActive: active });
    }
  }

  render() {
    const { data, store } = this.props;
    const { locationMapActive } = this.state;
    const { ui: { mediaLarge } } = store;

    let mapMode: MapMode;
    if (locationMapActive) {
      mapMode = 'picker';
    } else if (this.mapActivationRatio === 1.0) {
      mapMode = 'requests';
    } else {
      mapMode = 'inactive';
    }

    return (
      <div>
        <Nav activeSection="report" />

        {/* Outer box needs to take up at least the screen size on desktop, so
            that the content can center over the map and keep the footer from
            encroaching up. */}
        <div className={`mn mn--full mn--nv-s ${CONTAINER_STYLE.toString()}`} style={{ backgroundColor: 'transparent' }}>
          { mediaLarge &&
            <div className={BACKGROUND_MAP_CONTAINER_STYLE}>
              <LocationMap
                store={store}
                loopbackGraphql={this.loopbackGraphql}
                mode={mapMode}
                opacityRatio={this.mapActivationRatio}
              />
            </div>
          }

          { /* We can't put any wrappers around the dialogs below because in the
               case of RequestDialog's location picker, it needs to make sure nothing
               is overlapping the map so it can get clicks. */}

          { data.view === 'home' &&
            <HomeDialog
              store={store}
              loopbackGraphql={this.loopbackGraphql}
              stage={data.stage}
            /> }

          { data.view === 'request' &&
            <RequestDialog
              store={store}
              stage={data.stage}
              loopbackGraphql={this.loopbackGraphql}
              routeToServiceForm={this.routeToServiceForm}
              setLocationMapActive={this.setLocationMapActive}
            />}
        </div>

      </div>
    );
  }
}
