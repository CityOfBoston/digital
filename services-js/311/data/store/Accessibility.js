// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
