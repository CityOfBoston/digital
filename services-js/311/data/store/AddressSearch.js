// @flow

import { observable, action, computed, runInAction } from 'mobx';

import type { LoopbackGraphql } from '../dao/loopback-graphql';
import searchAddress from '../dao/search-address';
import reverseGeocode from '../dao/reverse-geocode';
import type { AddressUnit, SearchAddressPlace } from '../types';

// This class co-ordinates location queries between the LocationPopUp pane
// and the LocationMap map. LocationPopUp automatically updates the
// RequestForm (which contains the data that will go into the request) based
// on changes to this instance.
export default class AddressSearch {
  @observable query: string = '';
  @observable searching: boolean = false;
  @observable lastQuery: string = '';
  @observable.ref lastSearchError: ?Error = null;

  @observable.shallow places: ?Array<SearchAddressPlace> = null;
  @observable _currentPlaceIndex: number = -1;
  @observable highlightedPlaceIndex: number = -1;
  @observable _currentUnitIndex: number = 0;
  @observable
  currentReverseGeocodeLocation: ?{ lat: number, lng: number } = null;
  @observable currentReverseGeocodeLocationIsValid: boolean = true;

  @observable mode: 'search' | 'geocode' = 'search';

  // We keep track of whether the constituent seems to be pointing at a location
  // on the map or choosing from the location menu, since we have to choose to
  // send either lat/lng *or* address string to Open311. lat/lng matches the map
  // exactly, but only address string captures the unit.
  @observable intent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';

  // Used so that LocationMap doesn't zoom things under the search box.
  searchPopupWidth: number = 0;

  loopbackGraphql: ?LoopbackGraphql;

  start(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;
  }

  stop() {
    this.loopbackGraphql = null;
  }

  @computed
  get currentPlaceIndex(): number {
    return this._currentPlaceIndex;
  }

  set currentPlaceIndex(currentPlaceIndex: number) {
    this._currentPlaceIndex = currentPlaceIndex;
    this.intent = 'ADDRESS';
  }

  @computed
  get currentUnitIndex(): number {
    return this._currentUnitIndex;
  }

  set currentUnitIndex(currentUnitIndex: number) {
    this._currentUnitIndex = currentUnitIndex;
    this.intent = 'ADDRESS';
  }

  @action
  async search(selectFirst: boolean): Promise<void> {
    if (!this.loopbackGraphql) {
      return;
    }

    this.places = null;
    this.lastQuery = this.query;
    this.lastSearchError = null;
    this.currentReverseGeocodeLocation = null;
    this.mode = 'search';

    if (!this.query) {
      return;
    }

    try {
      this.searching = true;
      const places = await searchAddress(this.loopbackGraphql, this.query);

      runInAction('search - searchAddress success', () => {
        this.searching = false;
        this.places = places;
        this.currentPlaceIndex = places.length === 1 || selectFirst ? 0 : -1;
        this.highlightedPlaceIndex = this.currentPlaceIndex;
        this.currentUnitIndex = 0;
        this.intent = 'ADDRESS';
      });
    } catch (err) {
      runInAction('search - searchAddress error', () => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat &&
          !err.server &&
          window._opbeat('captureException', err);
      });
    }
  }

  @computed
  get location(): ?{ lat: number, lng: number } {
    return (
      this.currentReverseGeocodeLocation ||
      (this.currentPlace ? this.currentPlace.location : null)
    );
  }

  set location(location: {| lat: number, lng: number |}) {
    this.query = '';

    if (
      this.location &&
      this.location.lat === location.lat &&
      this.location.lng === location.lng
    ) {
      return;
    }

    if (!this.loopbackGraphql) {
      return;
    }

    this.currentReverseGeocodeLocation = location;
    this.places = null;
    this.mode = 'geocode';
    this.lastQuery = '';
    this.lastSearchError = null;

    this.searching = true;
    reverseGeocode(this.loopbackGraphql, location).then(
      action('reverseGeocode success', place => {
        this.searching = false;

        if (place) {
          // We overwrite the location in the place to stay with the one picked so
          // that the map marker doesnâ€™t move.
          this.places = [
            {
              location,
              address: place.address,
              addressId: place.addressId,
              units: place.units,
            },
          ];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = true;
          this.intent = 'LATLNG';
        } else {
          this.places = [];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = false;
        }
      }),
      action('reverseGeocode error', err => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat && window._opbeat('captureException', err);
      })
    );
  }

  @computed
  get currentPlace(): ?SearchAddressPlace {
    return (
      (this.places &&
        this.places.length > this.currentPlaceIndex &&
        this.places[this.currentPlaceIndex]) ||
      null
    );
  }

  @computed
  get currentUnit(): ?AddressUnit {
    return (
      (this.currentPlace &&
        this.currentPlace.units.length > this.currentUnitIndex &&
        this.currentPlace.units[this.currentUnitIndex]) ||
      null
    );
  }

  @computed
  get notFound(): boolean {
    return !!this.places && this.places.length === 0;
  }

  @computed
  get address(): string {
    if (this.currentUnit) {
      return this.currentUnit.address;
    } else if (this.currentPlace) {
      return this.currentPlace.address;
    } else {
      return '';
    }
  }

  @computed
  get addressId(): ?string {
    if (this.currentUnit) {
      return this.currentUnit.addressId;
    } else if (this.currentPlace) {
      return this.currentPlace.addressId;
    } else {
      return null;
    }
  }

  @computed
  get units(): Array<AddressUnit> {
    return this.currentPlace ? this.currentPlace.units : [];
  }
}
