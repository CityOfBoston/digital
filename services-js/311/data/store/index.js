// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
