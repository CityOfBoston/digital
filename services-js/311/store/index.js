// @flow
/* global window */
/* eslint no-underscore-dangle: 0 */

import { createStore, compose, applyMiddleware } from 'redux';
import type { Store as ReduxStore } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';
import inject from 'redux-inject';

import type { RequestAdditions } from '../server/next-handlers';
import { reducer } from './modules';
import type { AppState, AppAction } from './modules';
import makeLoopbackGraphql from './deps/loopback-graphql';
import type { LoopbackGraphql } from './deps/loopback-graphql';

// Re-export these to keep things centralized in the “store” module.
export type State = AppState;
export type Action = AppAction;

export type Store = ReduxStore<State, Action>;

export type Deps = {|
  loopbackGraphql: LoopbackGraphql,
|}

// We implement our own Dispatch type to account for the actions and return
// values provided by our middleware. (The Redux built-in type is from Action
// (which must have a type property) to Action. We need to take Actions that
// don't have properties and return pretty much anything, since that’s what
// thunks can return.
export type Dispatch = (action: Action) => any;

// For the browser, we keep a singleton store that can be re-used across pages.
// A module global is the only mechanism for sharing between Next.js pages.
let browserStore = null;

// Factored out to allow for injecting test dependencies and state while
// maintaining the same middleware.
export function makeStore(deps: ?Deps = null, initialState: ?State = null): Store {
  // Support for Redux Devtools Extension: https://github.com/zalmoxisus/redux-devtools-extension
  const composeEnhancers = (process.browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(
      inject(deps),
      thunk,
      reduxPackMiddleware,
    ),
  );

  if (initialState) {
    return createStore(reducer, initialState, enhancers);
  } else {
    return createStore(reducer, enhancers);
  }
}

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 *
 * req is the Node request (expected to contain injected dependencies for server
 * data fetching) that should be provided when creating the store for the
 * server’s getInitialProps calls. Can be null when server rendering or on the
 * client.
 */
export default function getStore(req: ?RequestAdditions, initialState: ?State = null): Store {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const deps = {
    loopbackGraphql: makeLoopbackGraphql(req),
  };

  const store = makeStore(deps, initialState);

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
