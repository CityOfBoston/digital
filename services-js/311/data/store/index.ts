/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
