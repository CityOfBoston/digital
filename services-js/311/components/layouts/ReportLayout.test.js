// @flow
/* eslint no-fallthrough: 0 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import ReportLayout from './ReportLayout';
import type { Service, ServiceSummary } from '../../data/types';
import { makeStore } from '../../data/store';

const MOCK_SERVICE_SUMMARIES: ServiceSummary[] = [{
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  hasMetadata: true,
  locationRequired: true,
}];

const MOCK_SERVICE: Service = {
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  hasMetadata: true,
  metadata: {
    attributes: [{
      required: false,
      type: 'TEXT',
      code: 'ST-CMTS',
      description: 'Please provide any other relevant information:',
      values: null,
    }, {
      required: false,
      type: 'STRING',
      code: 'INFO-CSIRMV1',
      description: '**All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**',
      values: null,
    }, {
      required: true,
      type: 'SINGLEVALUELIST',
      code: 'SR-CSIRMV1',
      description: 'How many dimensions have been breached?',
      values: [{ key: 'One', name: 'One' }, { key: 'Two', name: 'Two' }, { key: 'Three', name: 'Three' }, { key: 'More than Three', name: 'More than Three' }],
    }],
  },
};

let store;
let defaultProps;

beforeEach(() => {
  store = makeStore();

  defaultProps = {
    showServiceForm: jest.fn(),
    loopbackGraphql: jest.fn(),
    goToReportForm: jest.fn(),
  };
});

describe('report form', () => {
  let initialProps;
  beforeEach(() => {
    initialProps = {
      view: 'summaries',
      serviceSummaries: MOCK_SERVICE_SUMMARIES,
      pickLocation: false,
    };
  });

  test('title', () => {
    expect(ReportLayout.getTitle(initialProps)).toEqual('Report a Problem');
  });

  test('rendering', () => {
    const component = renderer.create(
      <Provider store={store}>
        <ReportLayout initialProps={initialProps} {...defaultProps} />
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('service form', () => {
  describe('existing', () => {
    let initialProps;

    beforeEach(() => {
      initialProps = {
        view: 'service',
        code: 'CSMCINC',
        service: MOCK_SERVICE,
        pickLocation: false,
      };
    });

    test('title', () => {
      expect(ReportLayout.getTitle(initialProps)).toEqual('Cosmic Incursion');
    });

    test('rendering', () => {
      const component = renderer.create(
        <Provider store={store}>
          <ReportLayout initialProps={initialProps} {...defaultProps} />
        </Provider>,
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('missing', () => {
    let initialProps;

    beforeEach(() => {
      initialProps = {
        view: 'service',
        code: 'INFEARTHS',
        service: null,
        pickLocation: false,
      };
    });

    test('title', () => {
      expect(ReportLayout.getTitle(initialProps)).toEqual('Not found');
    });

    test('rendering missing', () => {
      const component = renderer.create(
        <Provider store={store}>
          <ReportLayout initialProps={initialProps} {...defaultProps} />
        </Provider>,
    );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});

describe('location picker', () => {
  let initialProps;

  beforeEach(() => {
    initialProps = {
      view: 'service',
      code: 'CSMCINC',
      service: MOCK_SERVICE,
      pickLocation: true,
    };
  });

  test('title', () => {
    expect(ReportLayout.getTitle(initialProps)).toEqual('Choose location');
  });

  test('rendering', () => {
    const component = renderer.create(
      <Provider store={store}>
        <ReportLayout initialProps={initialProps} {...defaultProps} />
      </Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
