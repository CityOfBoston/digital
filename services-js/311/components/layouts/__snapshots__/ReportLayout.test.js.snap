exports[`location picker rendering 1`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
</div>
`;

exports[`report form rendering 1`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1t5zf7p": "",
      }
    }>
    <div
      className="sh sh--y"
      style={
        Object {
          "marginBottom": 40,
        }
      }>
      <h2
        className="sh-title">
        311: Boston City Services
      </h2>
    </div>
    <div
      className={
        Object {
          "data-css-9sglvw": "",
        }
      }>
      <textarea
        className={
          Object {
            "data-css-1j8x7ul": "",
          }
        }
        defaultValue=""
        name="description"
        onInput={[Function]}
        placeholder="How can we help?" />
      <ul
        className={
          Object {
            "data-css-117fchh": "",
          }
        }>
        <li
          className={
            Object {
              "data-css-14nloni": "",
            }
          }>
          <a
            className={
              Object {
                "data-css-3br7c8": "",
              }
            }
            href="/report/CSMCINC/location"
            onClick={[Function]}>
            Cosmic Incursion
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`service form existing rendering 1`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1t5zf7p": "",
      }
    }>
    <div
      className="sh sh--y"
      style={
        Object {
          "marginBottom": 40,
        }
      }>
      <h2
        className="sh-title">
        Cosmic Incursion
      </h2>
    </div>
    <div
      className={
        Object {
          "data-css-dgukzt": "",
        }
      }>
      
    </div>
    <div>
      <label>
        <p>
          Please provide any other relevant information:
           
        </p>
        <textarea
          className={
            Object {
              "data-css-1bm19ha": "",
            }
          }
          name="ST-CMTS"
          onInput={[Function]} />
      </label>
      <p>
        **All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**
      </p>
      <label>
        <p>
          How many dimensions have been breached?
        </p>
        <select
          name="SR-CSIRMV1"
          onInput={[Function]}>
          <option
            value="One">
            One
          </option>
          <option
            value="Two">
            Two
          </option>
          <option
            value="Three">
            Three
          </option>
          <option
            value="More than Three">
            More than Three
          </option>
        </select>
      </label>
    </div>
    <div>
      <h2
        className={
          Object {
            "data-css-16pm9zd": "",
          }
        }>
        CONTACT FORM
      </h2>
      <div
        className={
          Object {
            "data-css-1w0rge6": "",
          }
        }>
        (will not be shared with public; leave blank to submit anonymously)
      </div>
      <label
        className={
          Object {
            "data-css-12a7iqv": "",
          }
        }>
        <span>
          First Name (required)
        </span>
        <input
          className={
            Object {
              "data-css-b35x0h": "",
            }
          }
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-12a7iqv": "",
          }
        }>
        <span>
          Last Name (required)
        </span>
        <input
          className={
            Object {
              "data-css-b35x0h": "",
            }
          }
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-12a7iqv": "",
          }
        }>
        <span>
          Email (required)
        </span>
        <input
          className={
            Object {
              "data-css-b35x0h": "",
            }
          }
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-12a7iqv": "",
          }
        }>
        <span>
          Phone
        </span>
        <input
          className={
            Object {
              "data-css-b35x0h": "",
            }
          }
          name="phone"
          onChange={[Function]}
          placeholder="Phone"
          type="tel"
          value="" />
      </label>
    </div>
    <button
      disabled={true}
      onClick={[Function]}>
      Submit
    </button>
  </div>
</div>
`;

exports[`service form missing rendering missing 1`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1t5zf7p": "",
      }
    }>
    <div
      className="sh sh--y"
      style={
        Object {
          "marginBottom": 40,
        }
      }>
      <h2
        className="sh-title">
        Service not found
      </h2>
    </div>
  </div>
</div>
`;
